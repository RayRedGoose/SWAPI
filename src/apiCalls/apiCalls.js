export const getMoviesData = () => {
  return fetch('https://swapi.co/api/films/')
    .then(response => {
      if (!response.ok) {
        throw Error('Movies wasn\'t fetched')
      }
      return response.json()
    })
    .then(data => data.results.sort((a, b) => {
      return a.episode_id - b.episode_id
    }))
}

export const getAnyData = (url, type) => {
  return fetch(url)
    .then(response => {
      if (!response.ok) {
        throw Error(`${type} wasn\'t fetched`)
      }
      return response.json()
    })
}

// const getCharacterData = (array) => {
//   return array.map((character, index) => {
//     return fetch(character)
//     .then(response => response.json())
//   })
// }
//
// const makePromises = Promise.all(getCharacterData)
//   .then(promises => promises.map(promise => {
//
//     const characterName = promise.name;
//
//     return Promise.all([speciesFetch(promise), planetFetch(promise), filmsFetch(promise)])
//     .then(info => {return {name: characterName, species: info[0], planet: info[1].planetName, population: info[1].planetPopulation, films: info[2]}})
//     .then(characterStats => {this.setState({characters: [...this.state.characters, characterStats]})})
//   }))
//
//   const speciesFetch = (promise) => {
//     fetch(promise.species)
//     .then(res => res.json())
//     .then(data => data.name)
//   }
//
//   const planetFetch = (promise) => {
//     fetch(promise.homeworld)
//     .then(res => res.json())
//     .then(data => {return {planetName: data.name, planetPopulation: data.population}})
//   }
//
//   const filmsFetch = (promise) => {
//     const filmsPromises = promise.films.map(film => {
//       return fetch(film)
//       .then(res => res.json())
//       .then(data => data.title)
//
//       return Promise.all(filmsPromises)
//       .then(data => {return data})
//     })
//   }
