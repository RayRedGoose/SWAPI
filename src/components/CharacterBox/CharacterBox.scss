.character-box {
  display: flex;
  flex-wrap: wrap;
  justify-content: flex-start;
  margin: 20px 20px 80px;
  padding: 20px;
  position: relative;
  min-height: 300px;
}

.button-set {
  width: 50%;
  position: absolute;
  bottom: -20px; left: 40%;
  button {
    background-color: rgba(255, 255, 255, 0.9);
    border: none;
    color: #5A2C05;
    cursor: pointer;
    font-size: 18px;
    text-transform: lowercase;
    padding: 0.4em 2em;
    margin-right: 20px;
    &:hover {
      background: #5A2C05;
      color: #F1AA61;
    }
  }
}
