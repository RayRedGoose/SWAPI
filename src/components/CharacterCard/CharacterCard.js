import React from 'react';
import './CharacterCard.scss';
import FavoriteButton from '../FavoriteButton/FavoriteButton'

const CharacterCard = ({name, species, planet, population, films, addFavoriteCard}) => {

  const filmList = () => {
    films.map(film => {
      return <li>{film}</li>
    })
  }

  return (
    <section className="character-card">
      <header>
        <h3>{name}</h3>
        <p>{species}</p>
      </header>
      <section className="character-card-planet">
      <div>
        <p>{planet}:</p>
        <p>{population}</p>
      </div>
      <ul>
        <h4 className="character-movies-header">Movies:</h4>
        {filmList}
      </ul>
      </section>
      <footer>
        <FavoriteButton card={name, species, planet, population, films} addFavoriteCard={addFavoriteCard} /> 
      </footer>
    </section>
  )
}

export default CharacterCard;
