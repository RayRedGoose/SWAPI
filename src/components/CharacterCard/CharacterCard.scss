.character-card {
  background-color: rgba(255, 255, 255, 0.8);
  font-size: 18px;
  margin-bottom: 20px;
  margin-right: 17px;
  padding: 20px;
  position: relative;
  width: 20%;
  header {
    border-bottom: 2px solid #5A2C05;
    text-align: center;
    width: 80%;
    margin: auto;
    h3 {
      color: #5A2C05;
      font-size: 24px;
      min-height: 60px;
    }
    p {
      color: rgba(90, 44, 5, 0.5);
      text-transform: lowercase;
      padding: 10px;
    }
  }
  footer {
    text-align: center;
  }
}

.character-card-planet {
  div {
    display: flex;
    justify-content: space-between;
    padding: 20px 0;
    p {
      color: #5A2C05;
      text-transform: uppercase;
      &:first-child {
        margin-left: 25px;
        &::before {
          content: '';
          width: 20px; height: 20px;
          background-image: url('../../images/mars.svg');
          position: absolute;
          left: 20px;
        }
      }
    }
  }
  ul {
    h4 {
      color: #5A2C05;
      border-bottom: 2px solid #5A2C05;
      padding-bottom: 5px;
      width: 40%;
      margin-bottom: 20px;
    }
    li {
      list-style-type: none;
      margin-left: 40px;
      margin-bottom: 15px;
      &:before {
        content: '';
        width: 20px; height: 20px;
        background-image: url('../../images/st.svg');
        position: absolute;
        left: 30px;
      }
    }
  }
}
