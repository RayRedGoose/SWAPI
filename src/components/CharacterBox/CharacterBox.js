import React, { Component } from 'react';
import CharacterCard from '../CharacterCard/CharacterCard';
import './CharacterBox.scss';
import FavoriteCharacters from '../FavoriteCharacters/FavoriteCharacters'

export default class CharacterBox extends Component {
  constructor() {
    super()
    this.state = {
      characters: []
    }
  }
  
  componentDidMount = () => {
    const limitedCharacters = this.props.characters.slice(0, 10)

    const characterPromises = limitedCharacters.map(character => {
      return fetch(character)
      .then(response => response.json())
      .catch(err => {
        this.setState({characters: [...this.state.characters, character]})
      })
    })

    const characterFacts = Promise.all(characterPromises)
      .then(promises => promises.map(promise => {

        const characterName = promise.name;

        const speciesFetch =
        fetch(promise.species)
        .then(res => res.json())
        .then(data => data.name)

        const planetFetch =
        fetch(promise.homeworld)
        .then(res => res.json())
        .then(data => {return {planetName: data.name, planetPopulation: data.population}})

        const filmsPromises =
        promise.films.map(film => {
          fetch(film)
          .then(res => res.json())
          .then(data => data.title)
        })

        const filmsFetch =
        Promise.all(filmsPromises)

        const characterInfo =
        Promise.all([speciesFetch, planetFetch, filmsFetch])
        .then(info => {return {name: characterName, species: info[0], planet: info[1].planetName, population: info[1].planetPopulation, films: info[2]}})
        .then(characterStats => {this.setState({characters: [...this.state.characters, characterStats]})})
      }))
    }


  generateCharacters = () => {
    return this.state.characters.map(character => {
      return (
      <CharacterCard
      key={Date.now() + Math.random()}
      name={character.name}
      species={character.species}
      planet={character.planet}
      population={character.population}
      films={character.films}
      addFavoriteCard={this.addFavoriteCard}
      />
    )
    })
  }

  addFavoriteCard = (card) => {
    this.setState({...this.state, favorite: [...this.state.favorite, card]})
  }

  render() {
    return (
      <section className="character-box">
        {
          //<FavoriteCharacters favorite={this.state.favorite} />
          // NOTE: this is `/favorite` block with favorite cards
        }
        {this.generateCharacters()}
      </section>
    )
  }
}
