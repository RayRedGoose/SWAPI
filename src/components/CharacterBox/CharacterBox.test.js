import React from 'react'
import characterResults from '../../mockData/fakeCharacters.js'
import movieResults from '../../mockData/fakeMovies.js'
import { shallow } from 'enzyme'
import CharacterBox from './CharacterBox'
import CharacterCard from '../CharacterCard/CharacterCard'
import { getAnyData } from '../../apiCalls/apiCalls'

jest.mock('../../apiCalls/apiCalls')

describe('CharacterBox', () => {
  let wrapper, characters, instance

  beforeEach(() => {
    characters = movieResults.map(movie => movie.characters)

    wrapper = shallow(
      < CharacterBox
        favorites={[characterResults[1]]}
        characters={['C-3PO']}
        addFavorite={jest.fn()}
        removeFavorite={jest.fn()}
      />)
    instance = wrapper.instance()
  })

  it('should render CharacterBox with all data passed in correctly', () => {
    expect(wrapper).toMatchSnapshot()
  })

  it('should call generateCharacters after rendering', () => {
    const spy = jest.spyOn(instance, 'generateCharacters').mockImplementation(arg => arg)
    instance.forceUpdate()

    expect(spy).toHaveBeenCalled()
  })

  it('should call addCharacters after rendering', async () => {
    const spy = jest.spyOn(instance, 'addCharacters').mockImplementation(() => {
      return Promise.resolve([{ name: 'C-3PO' }])
    })

    await instance.componentDidMount()

    expect(spy).toHaveBeenCalled()
  })

  it('should call getAnyData in all nested functions to fetch api data', () => {
    getAnyData.mockImplementation(() => {
      return Promise.resolve([{ name: 'C-3PO' }])
    })

    instance.addCharacters()

    expect(getAnyData).toHaveBeenCalledTimes(5)
  })

  it('should call makePromises as a nested fetch for addCharacters method', async () => {
    const spy = jest.spyOn(instance, 'makePromises').mockImplementation(() => {
      return Promise.resolve([{ name: 'C-3PO' }])
    })

    await instance.addCharacters()

    expect(spy).toHaveBeenCalled()
  })

  it('should call nestedCharacterFetch as a nested fetch for makePromises method', async () => {
    const mockSpecies = async () => {
      return {species: 'human'};
    }

    const spy = jest.spyOn(instance, 'nestedCharacterFetch').mockImplementation(() => {
      return Promise.resolve([{ name: 'C-3PO' }])
    })

    await instance.makePromises([mockSpecies])

    expect(spy).toHaveBeenCalled()
  })

  describe("nestedCharacterFetch method", () => {
    let mockCharacter

    beforeEach(() => {
      mockCharacter = {
        name: 'C-3PO',
        species: '/linkToSpecies',
        films: ['/linkToFilmOne'],
        homeworld: '/linkToHomeworld'
      }
    })

    it('should call speciesFetch as a nested fetch', async () => {
      const spy = jest.spyOn(instance, 'speciesFetch').mockImplementation(() => {
        return Promise.resolve([{ species: 'human' }])
      })

      await instance.nestedCharacterFetch(mockCharacter)

      expect(spy).toHaveBeenCalled()
    })

    it('should call planetFetch as a nested fetch', async () => {
      const spy = jest.spyOn(instance, 'planetFetch').mockImplementation(() => {
        return Promise.resolve([{ species: 'human' }])
      })

      await instance.nestedCharacterFetch(mockCharacter)

      expect(spy).toHaveBeenCalled()
    })

    it('should call filmsFetch as a nested fetch', async () => {
      const spy = jest.spyOn(instance, 'filmsFetch').mockImplementation(() => {
        return Promise.resolve([{ species: 'human' }])
      })

      await instance.nestedCharacterFetch(mockCharacter)

      expect(spy).toHaveBeenCalled()
    })
  })
})
