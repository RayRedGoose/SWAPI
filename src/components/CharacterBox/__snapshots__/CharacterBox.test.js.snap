// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CharacterBox should render CharacterBox if data is loading 1`] = `
<section
  className="character-box"
>
  <ArtooLoading />
  <div
    className="button-set"
  />
</section>
`;

exports[`CharacterBox should render CharacterBox with all data passed in correctly 1`] = `
<section
  className="character-box"
>
  <div
    className="button-set"
  />
</section>
`;

exports[`CharacterBox should render error if data was not fetched 1`] = `<C3POCard />`;
